src/hooks/useFetch/index.ts

```diff
@@ -0,0 +1,60 @@
+import { useState, useEffect, useReducer, Reducer } from "react";
+import axios from "axios";
+import { DEFAULT_REQUEST_OPTION } from "../../config";
+import APIReducer, { StateProps, ActionProps } from "./reducer";
+import {
+    createloadAction,
+    createSuccessAction,
+    createFailAction
+} from "./actionCreator";
+
+enum Method {
+    POST = "post",
+    GET = "get",
+    PATCH = "patch",
+    PUT = "put",
+    DELETE = "delete"
+}
+
+interface FetchProps<T> {
+    method?: Method;
+    url?: string;
+    body?: T;
+}
+
+const API = axios.create(DEFAULT_REQUEST_OPTION);
+
+export default function useFetch<T>({
+    method = Method.GET,
+    url,
+    body
+}: FetchProps<T>) {
+    const [request, setRequest] = useState({ method, url, body });
+    const [state, dispatchFetchState] = useReducer<
+        Reducer<StateProps<T>, ActionProps<T>>
+    >(APIReducer, {
+        data: undefined,
+        loading: false,
+        error: false,
+        status: false
+    });
+
+    const doFetch = () => {
+        const { method, url, body } = request;
+        dispatchFetchState(createloadAction());
+        API({
+            method,
+            url,
+            data: body
+        })
+            .then(res => dispatchFetchState(createSuccessAction(res)))
+            .catch(error => dispatchFetchState(createFailAction(error)));
+    };
+
+    useEffect(() => {
+        if (!request.url) return;
+        doFetch();
+    }, [request]);
+
+    return [state, setRequest];
```

Comment:
useFetch에서 초기 값으로는 비어있는 값을 넣어주고, setRequest로 request를 변경시켜서 doFetch를 실행시켜서 state를 변경시켜주는 건가요?

Reply:
```js
const Component = () => {
const [{loading, data, error}, setRequest] = useFetch({ method: 'get' , url })
```

초기값은 사용할 때 인자로 넣어줍니다!
setRequest는 후에 다른 요청이 있을 때 request를 변경해서 사용하려고..
→ 댓글 수정 API → 댓글 삭제 API 
근데 요청이 두번 연속으로 가면 먼저 간 요청이 실패하는 에러가 아직 있네요 하하하하하하하하
어렵네여 어려워

Reply:
저도 타입스크립트 처음인데 너무 어려워여 어려워어어ㅓ어ㅓ어어어

Reply:
```suggestion
    const Component = () => {
        const [url1FetchState, doFetch1] = useFetch({ method: 'get' , url1 })
        const [url2FetchState, doFetch2] = useFetch({ method: 'get' , url2 })
    
        doFetch1.then(doFetch2).catch();
        ...
    }
```
setRequest대신 doFetch를 받아서 promise를 처리하는건 어떤가요?